layout.go

```diff
@@ -243,6 +263,183 @@ func (r *recordReader) guid() (v [16]byte) {
 	return
 }
 
+func (r *recordReader) methodDefSig() (v MethodDefSig) {
+	if r.err != nil {
+		return
+	}
+	// TODO: Split off a signature reader into its own type and split off common reading methods into another type. Having one reader for two distinct types of streams leads to a confusing API.
+	// Create new record reader with the same heap and layout as r (to read CodedIndex values) but
+	// point it at the signature data to read.
+	sigR := *r
+	sigR.data = r.blob()
+	if r.err != nil {
+		return
+	}
+
+	fail := func() bool {
+		r.err = sigR.err
+		return r.err != nil
+	}
+
+	firstByte := sigR.uint8()
+	if fail() {
+		return
+	}
+	kind := firstByte & 0xF
+	if kind > uint8(flags.SigKind_VARARG) {
+		r.err = fmt.Errorf("signature kind is not a method def signature: %v", kind)
+		return
+	}
+	thisiness := firstByte & 0xF0
+	v.HasThis = thisiness&uint8(flags.SigAttributes_HASTHIS) != 0
+	v.ExplicitThis = thisiness&uint8(flags.SigAttributes_EXPLICITTHIS) != 0
+	if thisiness&uint8(flags.SigAttributes_GENERIC) != 0 {
+		v.Generic = sigR.compressedUint32()
+		if fail() {
+			return
+		}
+	}
+	paramCount := sigR.compressedUint32()
+	if fail() {
+		return
+	}
+
+	v.RetType = sigR.retType()
+	if fail() {
+		return
+	}
+	for i := uint32(0); i < paramCount; i++ {
+		v.Param = append(v.Param, sigR.param())
+		if fail() {
+			return
+		}
+	}
+	return
+}
+
+func (r *recordReader) param() (v SigParam) {
+	if r.err != nil {
+		return
+	}
+	v.Type = r.decodeType()
+	switch v.Type.Kind {
+	case flags.ElementType_BYREF:
+		v.Kind = ParamKind_ByRef
+	case flags.ElementType_TYPEDBYREF:
+		v.Kind = ParamKind_TypedByRef
+	default:
+		v.Kind = ParamKind_ByValue
+	}
+	return
+}
+
+func (r *recordReader) retType() (v RetType) {
+	if r.err != nil {
+		return
+	}
+	v.Type = r.decodeType()
```

Comment:
A param signature can start with an array of `CustomMod` which is not taken into account here. Document a TODO or implement.